### ISSUE TITLE
bn node crash shortly after startup

### ISSUE THREAD

**Issue by thecodefactory** [libbitcoin/libbitcoin-node#470]
```
$ gdb ./bn
GNU gdb (Debian 12.1-4+b1) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./bn...
(gdb) run
Starting program: /mnt/ssd1/libbitcoin/master/install/bin/bn
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff55ff6c0 (LWP 562588)]
================= startup Thu Feb 16 16:35:09 2023 ==================
Using default configuration settings.
Press CTRL-C to stop the node.
Please wait while the node is starting...
[New Thread 0x7ffff4dfe6c0 (LWP 562589)]
Bypassed seeding because of sufficient (10000 of 50) address quantity.
Node is started.
Not configured for inbound connections.
[New Thread 0x7fffeffff6c0 (LWP 562590)]
Creating 10 connections 5 at a time.
Node is running.
Outbound [3.222.208.128:8333] protocol version (70016) agent: /Satoshi:22.0.0/
Channel count (0).
Outbound [51.81.245.33:8333] protocol version (70016) agent: /Satoshi:24.99.0/
Accepted (1000>556>481) addresses from [51.81.245.33:8333].
Outbound [13.209.125.83:8333] protocol version (70015) agent: /Satoshi:0.19.0.1/
Outbound [178.200.241.12:8333] protocol version (70016) agent: /Satoshi:22.0.0/
Outbound [185.150.160.209:8333] protocol version (70016) agent: /Satoshi:0.21.1/
Outbound [193.138.218.161:55900] protocol version (70016) agent: /Satoshi:24.0.1/
Accepted (1000>554>495) addresses from [193.138.218.161:55900].
Outbound [46.223.223.216:8333] protocol version (70016) agent: /Satoshi:23.0.0/
Outbound [121.140.205.84:8333] protocol version (70016) agent: /Satoshi:22.0.0/
Outbound [34.172.31.246:8333] protocol version (70016) agent: /Satoshi:22.0.0/
Accepted (1000>557>485) addresses from [46.223.223.216:8333].
Outbound [35.198.155.181:8333] protocol version (70016) agent: /Satoshi:0.21.1/
Outbound [73.82.99.12:8333] protocol version (70016) agent: /Satoshi:0.21.1/
Outbound [99.132.141.86:8333] protocol version (70016) agent: /Satoshi:23.0.0/Knots:20220529/
Accepted (988>668>592) addresses from [99.132.141.86:8333].
Accepted (1000>581>512) addresses from [121.140.205.84:8333].
Accepted (1000>545>480) addresses from [178.200.241.12:8333].
Outbound [77.9.100.15:8333] protocol version (70016) agent: /Satoshi:22.0.0/
Accepted (1>0>0) addresses from [77.9.100.15:8333].
Outbound [3.135.186.102:8333] protocol version (70016) agent: /Satoshi:0.21.1/
Outbound [71.201.9.208:8333] protocol version (70016) agent: /Satoshi:0.21.0/
Outbound [217.180.221.162:8333] protocol version (70016) agent: /Satoshi:24.0.1/
Accepted (1000>573>490) addresses from [217.180.221.162:8333].
Accepted (997>544>466) addresses from [35.198.155.181:8333].
Outbound [13.213.211.229:8333] protocol version (70016) agent: /Satoshi:0.21.1/
Outbound [96.236.128.66:8333] protocol version (70016) agent: /Satoshi:22.0.0/
Accepted (1>1>1) addresses from [46.223.223.216:8333].
Accepted (991>553>449) addresses from [71.201.9.208:8333].
Outbound [178.238.232.35:8333] protocol version (70015) agent: /Satoshi:0.20.0/
Accepted (1000>574>490) addresses from [77.9.100.15:8333].
Accepted (1>1>1) addresses from [121.140.205.84:8333].
Accepted (1>0>0) addresses from [99.132.141.86:8333].
Accepted (1000>584>482) addresses from [178.238.232.35:8333].

Thread 3 "bn" received signal SIGFPE, Arithmetic exception.
[Switching to Thread 0x7ffff4dfe6c0 (LWP 562589)]
0x0000555555658d4b in libbitcoin::network::hosts::save (this=0x555555756d88, hosts=std::vector of length 1, capacity 1 = {...}) at src/net/hosts.cpp:185
185         const auto step = std::max(usable / accept, one);
(gdb) 
```

Build is from today.  Last week's build was solid.

**Comment by evoskuil** [comment]
Hey, thanks!!

**Comment by evoskuil** [comment]
Looks like a change to randomized host storage allowed for division by zero (somewhat randomly).

**Comment by evoskuil** [comment]
https://github.com/libbitcoin/libbitcoin-network/pull/310/commits/b72d9e60464706ae4f5f294bfc1f80ac1498e366

**Comment by evoskuil** [comment]
It’s merged, with yesterday’s work. Thanks a lot for building and reporting. Saved me a lot of time.

**Comment by thecodefactory** [comment]
Nice -- thanks for the quick fix! Looks solid again :smile:

**Comment by evoskuil** [comment]
With a full address pool and a singleton address message there was a 50% div by zero probability. I had kept my pool at 100k for a while and due to filtering it was no longer filling. But I had recorded three unexplained halts, since this change was introduced. Hopefully that explains them.

I don’t get a crash dump from my Win32 console runs. And with our pervasive use of noexcept the stack unwind should be reduced to what you are seeing, just an abort. It is possible to compile with the NOEXCEPT macros disabled if necessary.
